src/YarpUtilities/src/RosPublisher.cpp

```diff
@@ -0,0 +1,482 @@
+        return false;
+    }
+        
+    if ( (jointList.size() != jointPositions.size()) || 
+         (jointList.size() != jointVelocities.size()) ||
+         (jointList.size() != jointEfforts.size()))
+    {
+        std::cerr << printPrefix << "Joint states size mismatch. Unable to publish joint states." << std::endl;
+        return false;
+    }
+        
+    pub.msg.header.seq++;
+    pub.msg.header.stamp = m_pimpl->getTimeStampFromYarp();
+    
+    pub.msg.name.clear();
+    pub.msg.position.clear();
+    pub.msg.velocity.clear();
+    pub.msg.effort.clear();
+    
+    for (std::size_t idx = 0; idx < jointList.size(); idx++)
+    {
+        pub.msg.name.emplace_back(jointList(idx));
+        pub.msg.position.emplace_back(jointPositions(idx));
+        pub.msg.velocity.emplace_back(jointVelocities(idx));
+        pub.msg.effort.emplace_back(jointEfforts(idx));
+    }
+    
+    pub.ptr->write(pub.msg);
+    return true;
+}
+
+bool RosPublisher::publishWrench(const std::string& frame, 
+                                 BipedalLocomotion::GenericContainer::Vector<const double>::Ref wrench6d)
+{
+    std::string_view printPrefix = "[RosPublisher::publishWrench] ";
+    
+    if (!m_pimpl->initialized)
+    {
+        std::cerr << printPrefix << "Please call initialize before publishing." << std::endl;
+        return false;
+    }
+    
+    if (m_pimpl->wrenchPublisherMap.find(frame) == m_pimpl->wrenchPublisherMap.end())
+    {
+        std::cerr << printPrefix << "Frame does not exist. Please add wrench publisher before publishing." << std::endl;
+        return false;
+    }
+            
+    auto& pub = m_pimpl->wrenchPublisherMap.at(frame);    
+    if (pub.ptr == nullptr)
+    {
+        std::cerr << printPrefix << "Wrench publisher was not configured. Unable to publish wrench stamped messages." << std::endl;
+        return false;
+    }
+        
+    if (wrench6d.size() != 6)
+    {
+        std::cerr << printPrefix << "Expecting a 6d input vector. Unable to publish wrench." << std::endl;
+        return false;
+    }
+    
+    pub.msg.header.seq++;
+    pub.msg.header.frame_id = frame;
+    pub.msg.header.stamp = m_pimpl->getTimeStampFromYarp();
+    pub.msg.wrench.force.x = wrench6d(0);
+    pub.msg.wrench.force.y = wrench6d(1);
+    pub.msg.wrench.force.z = wrench6d(2);
+    
+    pub.msg.wrench.torque.x = wrench6d(3);
+    pub.msg.wrench.torque.y = wrench6d(4);
+    pub.msg.wrench.torque.z = wrench6d(5);
+    pub.ptr->write(pub.msg);
+    
+    return true;
+}
+
+yarp::rosmsg::TickTime RosPublisher::Impl::getTimeStampFromYarp()
+{
+    std::string_view printPrefix = "[RosPublisher::Impl::getTimeStampFromYarp] ";
+    yarp::rosmsg::TickTime rosTickTime;
+    double yarpTimeStamp = yarp::os::Time::now();
+
+    uint64_t time = static_cast<uint64_t>(yarpTimeStamp * 1000000000UL);
+    uint64_t nsec_part = time % 1000000000UL;
+    uint64_t sec_part = time / 1000000000UL;
```

Comment:
Can 't you cast directly ``yarpTimeStamp`` instead of multiplying it first times 1000000000UL?

Reply:
Actually this function is a copy-paste from `hde`. :D  
Although I remember some issues in the past with ROS timestamps specially while handling `/tf`, if we are not careful enough it throws a timestamp extrapolation error.

Maybe @Yeshasvitvs @lrapetti  could add some light on this?

Reply:
> Actually this function is a copy-paste from

@prashanthr05 I guess you got it from https://github.com/robotology/human-dynamics-estimation/blame/master/publishers/HumanStatePublisher/HumanStatePublisher.cpp#L517-L536

This part was added by @diegoferigo at the initial stage of HDE v2, may be he can add something to the conversation. However, I notice that it is similar to the idl autogenerated code for `TickTime.h` https://github.com/robotology/yarp/blob/514b62f0f3575437cd86001d113773cb16372727/src/libYARP_rosmsg/src/idl_generated_code/yarp/rosmsg/TickTime.h#L45-L53

CC @lrapetti

Reply:
I have no memory of the logic of this code after 2y but I guess that the conversion has to deal with the fact the yarp time is a double and ros time is a pair of unsigned (sec, nsec), and the conversion should maximize the range while preventing overflow.

I think that today you can do much better using chrono and converting first the double to ns and then from it extracting (sec, nsec).

Reply:
Thanks @Yeshasvitvs, @diegoferigo, after reading it a couple of times I realized why those operations were necessary. It is also true that if it required several people to guess what is going on, maybe it is worth a bit of refractory. Indeed, the hint of ``std::chrono`` may be a good move, https://en.cppreference.com/w/cpp/chrono/duration

Reply:
Changed to use `chrono` in https://github.com/dic-iit/bipedal-locomotion-framework/pull/156/commits/d57a286f4780f87c335ff0b3b91ca01443ffa069